File: backend/tourism-service/src/main/java/com/smarttouristplatform/tourismservice/service/ReviewService.java
package com.smarttouristplatform.tourismservice.service;

import com.smarttouristplatform.tourismservice.Exception.ResourceNotFoundException;
import com.smarttouristplatform.tourismservice.dto.AddReviewResponseRequest;
import com.smarttouristplatform.tourismservice.dto.CreateGuideReviewRequest;
import com.smarttouristplatform.tourismservice.dto.CreateHotelReviewRequest;
import com.smarttouristplatform.tourismservice.dto.UpdateReviewRequest;
import com.smarttouristplatform.tourismservice.model.CategoryRatings;
import com.smarttouristplatform.tourismservice.model.Review;
import com.smarttouristplatform.tourismservice.model.ReviewCategories;
import com.smarttouristplatform.tourismservice.repository.ReviewRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class ReviewService {

    @Autowired
    private ReviewRepository reviewRepository;

    public Review createGuideReview(String guideBookingId, CreateGuideReviewRequest request, String touristId) {
        Review review = new Review();
        BeanUtils.copyProperties(request, review);
        review.setTouristId(touristId);
        review.setGuideBookingId(guideBookingId);
        // In a real app, fetch guideId from guideBookingId
        review.setGuideId("sampleGuideId");
        review.setCreatedAt(LocalDateTime.now());
        review.setUpdatedAt(LocalDateTime.now());

        ReviewCategories categories = new ReviewCategories();
        BeanUtils.copyProperties(request.getCategories(), categories);
        review.setCategories(categories);

        return reviewRepository.save(review);
    }

    public Review createHotelReview(String hotelBookingId, CreateHotelReviewRequest request, String touristId) {
        Review review = new Review();
        BeanUtils.copyProperties(request, review);
        review.setTouristId(touristId);
        review.setHotelBookingId(hotelBookingId);
        // In a real app, fetch hotelId from hotelBookingId
        review.setHotelId("sampleHotelId");
        review.setCreatedAt(LocalDateTime.now());
        review.setUpdatedAt(LocalDateTime.now());

        CategoryRatings categoryRatings = new CategoryRatings();
        BeanUtils.copyProperties(request.getCategoryRatings(), categoryRatings);
        review.setCategoryRatings(categoryRatings);

        return reviewRepository.save(review);
    }

    public List<Review> getGuideReviews(String guideId, int page, int limit, String sort) {
        Sort sortBy = Sort.by(sort.startsWith("-") ? Sort.Direction.DESC : Sort.Direction.ASC, sort.replace("-", ""));
        Pageable pageable = PageRequest.of(page - 1, limit, sortBy);
        return reviewRepository.findByGuideId(guideId);
    }

    public List<Review> getHotelReviews(String hotelId, int page, int limit, String sort) {
        Sort sortBy = Sort.by(sort.startsWith("-") ? Sort.Direction.DESC : Sort.Direction.ASC, sort.replace("-", ""));
        Pageable pageable = PageRequest.of(page - 1, limit, sortBy);
        return reviewRepository.findByHotelId(hotelId);
    }

    public Review updateReview(String reviewId, UpdateReviewRequest request, String touristId) {
        Review existingReview = reviewRepository.findById(reviewId)
                .filter(review -> review.getTouristId().equals(touristId))
                .orElseThrow(() -> new ResourceNotFoundException("Review not found or access denied"));

        Optional.ofNullable(request.getRating()).ifPresent(existingReview::setRating);
        Optional.ofNullable(request.getComment()).ifPresent(existingReview::setComment);
        existingReview.setUpdatedAt(LocalDateTime.now());

        return reviewRepository.save(existingReview);
    }

    public void deleteReview(String reviewId, String touristId) {
        Review existingReview = reviewRepository.findById(reviewId)
                .filter(review -> review.getTouristId().equals(touristId))
                .orElseThrow(() -> new ResourceNotFoundException("Review not found or access denied"));
        reviewRepository.delete(existingReview);
    }

    public Review addReviewResponse(String reviewId, AddReviewResponseRequest request) {
        Review existingReview = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new ResourceNotFoundException("Review not found"));

        existingReview.setResponseText(request.getResponseText());
        existingReview.setUpdatedAt(LocalDateTime.now());
        return reviewRepository.save(existingReview);
    }
}

File: smart-tourist-platform (3)/client/src/services/review.service.js
import axiosInstance, { API_BASE } from '@/lib/axios';

const REVIEW_API = `${API_BASE}/reviews`;

export const reviewService = {
  createGuideReview: async (guideBookingId, rating, title, comment, categories) => {
    const response = await axiosInstance.post(`${REVIEW_API}/guides/${guideBookingId}`, {
      rating, title, comment, categories,
    });
    return response.data;
  },

  getGuideReviews: async (guideId) => {
    const response = await axiosInstance.get(`${REVIEW_API}/guides?guideId=${guideId}`);
    return response.data;
  },

  createHotelReview: async (hotelBookingId, rating, title, comment, categoryRatings) => {
    const response = await axiosInstance.post(`${REVIEW_API}/hotels/${hotelBookingId}`, {
      rating, title, comment, categoryRatings,
    });
    return response.data;
  },

  getHotelReviews: async (hotelId) => {
    const response = await axiosInstance.get(`${REVIEW_API}/hotels?hotelId=${hotelId}`);
    return response.data;
  },

  updateGuideReview: async (reviewId, data) => {
    const response = await axiosInstance.put(`${REVIEW_API}/guides/${reviewId}`, data);
    return response.data;
  },

  updateHotelReview: async (reviewId, data) => {
    const response = await axiosInstance.put(`${REVIEW_API}/hotels/${reviewId}`, data);
    return response.data;
  },

  deleteGuideReview: async (reviewId) => {
    await axiosInstance.delete(`${REVIEW_API}/guides/${reviewId}`);
  },

  deleteHotelReview: async (reviewId) => {
    await axiosInstance.delete(`${REVIEW_API}/hotels/${reviewId}`);
  },
};
